### PR TITLE
Texture rtt

### PR DESCRIPTION
**Pull request by mbouron**

### CODE REVIEW

File: libnopegl/src/node_rtt.c [hunk 1]
```diff
@@ -141,6 +142,17 @@ static int rtt_init(struct ngl_node *node)
...
             return NGL_ERROR_INVALID_ARG;
         }
+
+        const struct texture_priv *texture_priv = info.texture_priv;
+        struct texture_params *params = &texture_priv->params;
+        if (i == 0) {
+            s->width = params->width;
+            s->height = params->height;
+        } else if (s->width != params->width || s->height != params->height) {
+            LOG(ERROR, "all color texture dimensions do not match %dx%d != %dx%d",
```
Comment (ubitux) [inline libnopegl/src/node_rtt.c:156]: nit: add a `:` after "match"

File: libnopegl/src/node_rtt.c [hunk 2]
```diff
@@ -155,10 +167,30 @@ static int rtt_init(struct ngl_node *node)
...
+        params->usage |= NGLI_TEXTURE_USAGE_DEPTH_STENCIL_ATTACHMENT_BIT;
+        s->layout.depth_stencil.format = params->format;
+        s->layout.depth_stencil.resolve = o->samples > 1;
+    } else {
+        int depth_format = NGLI_FORMAT_UNDEFINED;
+        if (s->renderpass_info.features & NGLI_RENDERPASS_FEATURE_STENCIL)
+            depth_format = ngli_gpu_ctx_get_preferred_depth_stencil_format(gpu_ctx);
+        else if (s->renderpass_info.features & NGLI_RENDERPASS_FEATURE_DEPTH)
+            depth_format = ngli_gpu_ctx_get_preferred_depth_format(gpu_ctx);
+        s->layout.depth_stencil.format = depth_format;
```
Comment (ubitux) [inline libnopegl/src/node_rtt.c:204]: I think this belongs in another commit (this is currently in "rtt: move attachment dimensions check to rtt_init()")

File: libnopegl/src/node_rtt.c [hunk 3]
```diff
@@ -143,7 +153,7 @@ static int rtt_init(struct ngl_node *node)
             return NGL_ERROR_INVALID_ARG;
         }
 
-        const struct texture_priv *texture_priv = info.texture_priv;
+        struct texture_priv *texture_priv = info.texture_priv;
```
Comment (ubitux) [inline libnopegl/src/node_rtt.c]: Why do you need to change constness in this commit?

File: libnopegl/src/node_rtt.c [hunk 4]
```diff
@@ -168,7 +185,7 @@ static int rtt_init(struct ngl_node *node)
...
```
Comment (ubitux) [inline libnopegl/src/node_rtt.c:182]: ditto

File: libnopegl/src/rtt.c [hunk 5]
```diff
@@ -0,0 +1,299 @@
+/*
+ * Copyright 2023 Matthieu Bouron <matthieu.bouron@gmail.com>
+ * Copyright 2023 Nope Foundry
```
Comment (ubitux) [inline libnopegl/src/rtt.c:3]: this is derived work from `node_rtt.c` so you need to preserve "Copyright 2016-2022 GoPro Inc."

File: libnopegl/src/node_texture.c [hunk 6]
```diff
@@ -371,6 +372,43 @@ static int texture_prefetch(struct ngl_node *node)
...
+            .nb_interruptions = s->renderpass_info.nb_interruptions,
+            .nb_colors = 1,
+            .colors[0] = {
+                .attachment = s->texture,
+                .load_op = NGLI_LOAD_OP_CLEAR,
+                .store_op = NGLI_STORE_OP_STORE,
+                .clear_value[0] = o->clear_color[0],
+                .clear_value[1] = o->clear_color[1],
+                .clear_value[2] = o->clear_color[2],
+                .clear_value[3] = o->clear_color[3],
```
Comment (ubitux) [inline libnopegl/src/node_texture.c:398]: not possible to use `NGLI_ARG_VEC4()`?

File: libnopegl/src/node_texture.c [hunk 7]
```diff
@@ -371,6 +372,43 @@ static int texture_prefetch(struct ngl_node *node)
...
+    if (s->rtt) {
+        /* Transform the color textures coordinates so it matches how the
+         * graphics context uv coordinate system works regarding render targets */
+        ngli_gpu_ctx_get_rendertarget_uvcoord_matrix(gpu_ctx, s->image.coordinates_matrix);
+
+        int depth_format = NGLI_FORMAT_UNDEFINED;
+        if (s->renderpass_info.features & NGLI_RENDERPASS_FEATURE_STENCIL)
+            depth_format = ngli_gpu_ctx_get_preferred_depth_stencil_format(gpu_ctx);
+        else if (s->renderpass_info.features & NGLI_RENDERPASS_FEATURE_DEPTH)
+            depth_format = ngli_gpu_ctx_get_preferred_depth_format(gpu_ctx);
```
Comment (ubitux) [inline libnopegl/src/node_texture.c:384]: I feel like this is the 3rd time I'm seeing this chunk of code, you might want to consider a dedicated function somewhere

File: libnopegl/src/node_rtt.c (hunk 1 above)
Comment (mbouron) [reply libnopegl/src/node_rtt.c:156]: Fixed.

File: libnopegl/src/node_rtt.c (hunk 2 above)
Comment (mbouron) [reply libnopegl/src/node_rtt.c:204]: Yep. Moved to the next commit.

File: libnopegl/src/node_rtt.c (hunk 3 above)
Comment (mbouron) [reply libnopegl/src/node_rtt.c]: Because we need to write to the underlying texture params struct later on: `params->usage |= ...`

File: libnopegl/src/node_texture.c (hunk 6 above)
Comment (mbouron) [reply libnopegl/src/node_texture.c:398]: Nope. But this will change in another PR where I introduce a dedicated clear_color structure.

File: libnopegl/src/rtt.c (hunk 5 above)
Comment (mbouron) [reply libnopegl/src/rtt.c:3]: Fixed.

File: libnopegl/src/node_texture.c (hunk 7 above)
Comment (mbouron) [reply libnopegl/src/node_texture.c:384]: Will do in the next PR relative to render targets.

**Review by ubitux — APPROVED** [review]
